### PR TITLE
[oneDNN] Fix to #33282 , added support of X input broadcasting to oneDNN elementwise ops

### PR DESCRIPTION
**Pull request by jczaja**
### PR types
Bug fixes

### PR changes
OPs

### Describe
This is fix to #33282 . It does implement broadcasting for elementwise X input as previosuly only input Y was having broadcasting support. Inplace support for oneDNN elementwise kernels was disabled.

### CODE REVIEW

**Comment by lidanqing-vv** [comment]
@jczaja  Hi next time could you please name the PR with more specific description? For example this PR could be extend broadcast elementwise_add op with X(1 dims) Y(2 dims). 
It will be easier for release note

File: paddle/fluid/operators/elementwise/mkldnn/elementwise_mkldnn_op.h [hunk 1]
```diff
@@ -47,23 +47,13 @@ class EltwiseMKLDNNKernel : public framework::OpKernel<T> {
...
-        scale_x, scale_y, scale_o, key);
+        scale_x, scale_y, scale_o, ctx.OutputName("Out"));
 
     const auto src_x_memory = handler.AcquireSrcMemory(x);
     const auto src_y_memory = handler.AcquireSecondSrcMemory(y);
-
-    // For Inplace src and and dst are the same memory object
-    const auto dst_memory =
-        is_inplaced ? src_x_memory : handler.AcquireDstMemory(z);
+    const auto dst_memory = handler.AcquireDstMemory(z);
```
Comment (lidanqing-vv) [inline paddle/fluid/operators/elementwise/mkldnn/elementwise_mkldnn_op.h:57]: So we remove the inplace support for elementwise_add and elementwise_mul ? But you plan to remove it permanently or you will still add inplace later ?

File: paddle/fluid/platform/mkldnn_reuse.h [hunk 2]
```diff
@@ -674,15 +666,21 @@ class BinaryMKLDNNHandler : public platform::MKLDNNHandlerT<T, dnnl::binary> {
       const auto dst_tz =
           (z == nullptr) ? src_x_tz : framework::vectorize(z->dims());
```
Comment (lidanqing-vv) [inline paddle/fluid/platform/mkldnn_reuse.h]: if z is null, we point set dst_tz to src_x_tz, but what if src_x_tz is one dimension ? And why z could be nullptr, in which case z could be nullptr ?

File: paddle/fluid/platform/mkldnn_reuse.h [hunk 3]
```diff
@@ -674,15 +666,21 @@ class BinaryMKLDNNHandler : public platform::MKLDNNHandlerT<T, dnnl::binary> {
...
 
-      const auto src0_md = dnnl::memory::desc(
+      auto src0_md = dnnl::memory::desc(
           src_x_tz, platform::MKLDNNGetDataType<T>(), x->format());
       auto src1_md = dnnl::memory::desc(
           src_y_tz, platform::MKLDNNGetDataType<T>(), y->format());
-      if (rankdiff > 0) {
+      auto rankdiff = x->dims().size() - y->dims().size();
+      if (rankdiff > 0) {  // Second input is lesser than first
```
Comment (lidanqing-vv) [inline paddle/fluid/platform/mkldnn_reuse.h]: Sorry but why it is implemented this way.
I thought broadcast is copy values line by line until Y is same dimension as X dimension.
If X is 5*12, and Y is 5, rankdiff is 1?
or it is not like that ?

File: paddle/fluid/operators/elementwise/mkldnn/elementwise_mkldnn_op.h (hunk 1 above)
Comment (jczaja) [reply paddle/fluid/operators/elementwise/mkldnn/elementwise_mkldnn_op.h:57]: Inplace working with oneDNN caching is very hard , and there is little perf gain and some performance improvement. Among all of this elementwise ops for inplace are most difficult so I removed inplace support for elementwise oneDNN kernels.

File: paddle/fluid/platform/mkldnn_reuse.h (hunk 2 above)
Comment (jczaja) [reply paddle/fluid/platform/mkldnn_reuse.h]: Good observation.
1) z is null in situation when we want Binary primitive to write into allocation managed by oneDNN e.g. we do not want output be send to Tensor. This is only used in elementwise_mul_grad:77 as this operation consists of Binary + Reorder . Binary output is a temporary data that is an input to Reorder . Output of reorder goes into output tensor. 

2) Currently there in elementwise_mul_grad (src_x_tz) is always gradient of of Out buffer so I know it is never to be broadcasted. But I agree that code would be better if it works for z == nullptr and x being bradcasted. So I improve that. thanks

File: paddle/fluid/platform/mkldnn_reuse.h (hunk 3 above)
Comment (jczaja) [reply paddle/fluid/platform/mkldnn_reuse.h]: If X is [5,12] and Y is [5] then Y dims for oneDNN need to be [5,1] . And then oneDNN having inputs [5,12] and [5,1] will recognize that second input need broadcasting to have applied. So code you are referencing is just to change shape from [5] to [5,1]

**Review by lidanqing-vv — DISMISSED** [review]
LGTM

**Comment by jczaja** [comment]
@juncaipeng PR-CI-APPROVAL is complaining on wrong PADDLE_ENFORCE_EQ macro I touched in this PR and I do not understand what is wrong with it. couple you please advice?

**Comment by juncaipeng** [comment]
> @juncaipeng PR-CI-APPROVAL is complaining on wrong PADDLE_ENFORCE_EQ macro I touched in this PR and I do not understand what is wrong with it. couple you please advice?

The error infomation is `Possible errors include 1. the error message is empty / 2. the error message is too short / 3. the error type is not specified.`

The specification in https://github.com/PaddlePaddle/Paddle/wiki/Paddle-Error-Message-Writing-Specification-(English-Verison)

Maybe you can add more error messages ~

File: paddle/fluid/platform/mkldnn_reuse.h [hunk 4]
```diff
@@ -671,18 +663,24 @@ class BinaryMKLDNNHandler : public platform::MKLDNNHandlerT<T, dnnl::binary> {
...
+      const auto dst_tz = (z == nullptr) ? (rankdiff > 0 ? src_x_tz : src_y_tz)
+                                         : framework::vectorize(z->dims());
 
-      const auto src0_md = dnnl::memory::desc(
+      auto src0_md = dnnl::memory::desc(
           src_x_tz, platform::MKLDNNGetDataType<T>(), x->format());
       auto src1_md = dnnl::memory::desc(
           src_y_tz, platform::MKLDNNGetDataType<T>(), y->format());
-      if (rankdiff > 0) {
+      if (rankdiff > 0) {  // Second input is lesser than first
```
Comment (jakpiase) [inline paddle/fluid/platform/mkldnn_reuse.h]:
```suggestion
      if (rankdiff > 0) {  // Second input is smaller than first
```

File: paddle/fluid/platform/mkldnn_reuse.h [hunk 5]
```diff
@@ -644,14 +644,6 @@ class BinaryMKLDNNHandler : public platform::MKLDNNHandlerT<T, dnnl::binary> {
             platform::CreateKey(
                 dev_ctx, framework::vectorize(x->dims()), uniq_name,
                 (algo == dnnl::algorithm::binary_mul ? "M" : ""))) {
```
Comment (jakpiase) [inline paddle/fluid/platform/mkldnn_reuse.h]: Why aren't you just passing algo to CreateKey function? It can convert dnnl:algorithm datatype to a string

**Comment by lidanqing-vv** [comment]
@jczaja Please cherry-pick this PR into release/2.1 after this PR pass all CIs. Thanks.

**Comment by lidanqing-vv** [comment]
@jczaja  I saw you remove the whole test of inplace tests. Since 22 is tag day and Baidu expect today to merge this PR. Is it possible to keep minor change and make it pass CIs. If CI still fail not because of this PR, let me know.

**Comment by jczaja** [comment]
@juncaipeng I would like You to advice me on failing **test_pass_builder** . Reason of its failure is that
"<path>/tmp/test_viz_pass" file cannot be created , as this file is to be created/opened via std::ofstream , and in this
situation **tmp** directory does not exist so ofstream fials due to missing tmp directory. When this tmp directory is to be created?

Error message:

2021-06-18 19:51:35 Error Message Summary:
2021-06-18 19:51:35 ----------------------
2021-06-18 19:51:35 **UnavailableError: Can not open file /workspace/Paddle/build/python/paddle/fluid/tests/unittests/tmp/test_viz_pass for printing the graph**.
2021-06-18 19:51:35   [Hint: Expected fout->good() == true, but received fout->good():0 != true:1.] (at /workspace/Paddle/paddle/fluid/framework/ir/graph_viz_pass.cc:44)

**Review by arlesniak — APPROVED** [review]
LGTM

**Review by lidanqing-vv — APPROVED** [review]
LGTM

**Comment by jakpiase** [comment]
LGTM

**Review by jakpiase — APPROVED** [review]

**Pull request merged**